test/unit/test_grouped_loader.py

```diff
@@ -0,0 +1,238 @@
+        )
+        data_loader.close()
+
+        assert "we need at least two groups" in str(e_info.value)
+
+
+def test_sample_indices_intra():
+    """
+    Test the number of samples and indices are correct for a single image sample
+    """
+    data_dir_path = join(DataPaths["h5"], "train")
+    image_shape = (64, 64, 60)
+    common_args = dict(
+        file_loader=FileLoaderDict["h5"],
+        labeled=True,
+        sample_label="all",
+        intra_group_prob=0.5,
+        intra_group_option="forward",
+        sample_image_in_group=True,
+        seed=None,
+    )
+    data_loader = GroupedDataLoader(
+        data_dir_path=data_dir_path, image_shape=image_shape, **common_args
+    )
+
+    assert data_loader.sample_indices is None
+    assert data_loader._num_samples == 2
+
+
+def test_group_indices_inter():
+    """
+    Test the number of samples and indices are correct when entire group is sampled
+    """
+    data_dir_path = join(DataPaths["h5"], "train")
+    image_shape = (64, 64, 60)
+    common_args = dict(
+        file_loader=FileLoaderDict["h5"],
+        labeled=True,
+        sample_label="all",
+        intra_group_prob=0,
+        intra_group_option="forward",
+        sample_image_in_group=False,
+        seed=None,
+    )
+    data_loader = GroupedDataLoader(
+        data_dir_path=data_dir_path, image_shape=image_shape, **common_args
+    )
+
+    ni = np.array(data_loader.num_images_per_group)
+    num_samples = np.sum(ni) * (np.sum(ni) - 1) - sum(ni * (ni - 1))
+
+    sample_indices = data_loader.sample_indices
+    sample_indices.sort()
+    unique_indices = list(set(sample_indices))
+    unique_indices.sort()
+
+    assert data_loader._num_samples == num_samples
+    assert sample_indices == unique_indices
+
+
+def test_group_indices_intra():
+    """
+    Test sample numbers and indices are correct when entire group is sampled
+    """
+    data_dir_path = join(DataPaths["h5"], "train")
+    image_shape = (64, 64, 60)
+    for group_option in ["forward", "backward", "unconstrained"]:
+        common_args = dict(
+            file_loader=FileLoaderDict["h5"],
+            labeled=True,
+            sample_label="all",
+            intra_group_prob=1,
+            intra_group_option=group_option,
+            sample_image_in_group=False,
+            seed=None,
+        )
+        data_loader = GroupedDataLoader(
+            data_dir_path=data_dir_path, image_shape=image_shape, **common_args
+        )
+
+        ni = np.array(data_loader.num_images_per_group)
+        num_samples = sample_count(ni, direction=group_option)
+
+        sample_indices = data_loader.sample_indices
+        sample_indices.sort()
+        unique_indices = list(set(sample_indices))
+        unique_indices.sort()
+
+        assert data_loader._num_samples == num_samples
+        assert sample_indices == unique_indices
+
+
+def sample_count(ni, direction="forward"):
```

Comment:
I'm wondering leaving a default value might be dangerous

Reply:
Agreed, amended.